text/0008-uniform-random-tip-selection/0008-uniform-random-tip-selection.md

```diff
@@ -0,0 +1,154 @@
+```
+
+enum Score (
+    LAZY = 0
+    SEMI_LAZY = 1
+    NON_LAZY = 2
+)
+
+const (
+    C1 = 8
+    C2 = 13
+    M = 15
+)
+
+func score(tip Tip) Score {
+    
+    // if the LSMI to YMRSI delta is over C1, then the tip is lazy
+    if (LSMI - YMRSI(tip) > C1) {
+        return Score.LAZY
+    }
+    
+    // if the OMRSI to LSMI delta is over M/below-max-depth, then the tip is lazy
+    if (LSMI - OMRSI(tip) > M) {
+        return Score.LAZY
+    }
+    
+    if (LSMI - OMRSI(tip) > C2) {
+        return Score.SEMI_LAZY
+    }
+
+    return Score.NON_LAZY
+}
+```
+
+### Random Tip-Selection
+
+A node should keep a set of non-lazy tips (score 2).
+Every time a node is asked to select to tips to be approved it will just pick randomly from the set. 
+A node must not execute tip-selection if it is not synchronized.
+
+A tip should not be removed from the tips set immediately after it was selected in `select()`, 
+in order to make it possible for it to be re-selected, which in turn makes the Tangle wider
+and improves synchronization speed. A tip is removed from the tips set if `X` amount of direct
+approvers are reached or if a certain amount of time `T` passed. 
+It is recommended to use `X` = 2 and `T` = 3  but the threshold should be configurable.
+
+### Purpose Of Semi-Lazy Tips
+
+Semi-Lazy tips are not eligible for tip-selection, but the coordinator node may implement a tip selection algorithm
+that confirms semi-lazy tips. Semi-lazy tips will usually be left behind, but parties interested in having them confirmed
+are incentivized to run spammers that will actively reduce the amount of semi-lazy tips eligible for coordinator's tip selection. 
+Given a coordinator that chooses semi-lazy tips, running such spammers may get those messages confirmed before
+they become lazy.
+
+
+# Drawbacks
+
+Depending on when and how often `YMRSI`/`OMRSI` values are computed, this tip-selection could still
+have a slow runtime, as one would need to constantly walk down the Tangle in order to compute those
+values. However, smart caching might resolve this issue. 
+
+# Rationale and alternatives
+
+The previous tip-selection was written in accordance to the original IOTA whitepaper, as it also
+functioned as part of the consensus mechanism.
+However, relatively soon it became apparent that the cumulative weight computation was too heavy
+for an actual high throughput scenario and as such, the CW calculation is currently not used within
+node implementations at all.
+
+Because confirmations with the [white-flag](https://github.com/iotaledger/protocol-rfcs/blob/master/text/0005-white-flag/0005-white-flag.md) approach no longer only approve cones with state mutations
+which are consistent with a previous ledger state, it makes sense to alter the tip-selection to provide 
+a fast way to get tips to approve with one's own message.
+The only important thing is to disincentive lazy behaviour in order to be able to maximize confirmation rate.
```

Comment:
```suggestion
A node must not execute tip-selection if it is not synchronized.

A tip should not be removed from the tips set immediately after it was selected in `select()`,  to make it possible for it to be re-selected, which in turn makes the Tangle wider
and improves synchronization speed. A tip is removed from the tips set if `X` amount of direct
approvers are reached or if a certain amount of time `T` passed. 
It is recommended to use `X` = 2 and `T` = 3  but the threshold should be configurable.

### Purpose Of Semi-Lazy Tips

Semi-Lazy tips are not eligible for tip-selection, but the coordinator node may implement a tip selection algorithm
that confirms semi-lazy tips. Semi-lazy tips will usually be left behind, but parties interested in having them confirmed
are incentivized to run spammers that will actively reduce the amount of semi-lazy tips eligible for coordinator's tip selection. 
Given a coordinator that chooses semi-lazy tips, running such spammers may get those messages confirmed before
they become lazy.


# Drawbacks

Depending on when and how often `YMRSI`/`OMRSI` values are computed, this tip-selection could still
have a slow runtime, as one would need to constantly walk down the Tangle to compute those
values. However, smart caching might resolve this issue. 

# Rationale and alternatives

The previous tip-selection was written in accordance to the original IOTA whitepaper, as it also
functioned as part of the consensus mechanism.
However, relatively soon it became apparent that the cumulative weight computation was too heavy
for an actual high throughput scenario and, as such, the CW calculation is currently not used within
node implementations at all.

Because confirmations with the [white-flag](https://github.com/iotaledger/protocol-rfcs/blob/master/text/0005-white-flag/0005-white-flag.md) approach no longer approve cones only with state mutations,
which are consistent with a previous ledger state, it makes sense to alter the tip-selection to provide 
a fast way to get tips to approve with one's own message.
The only important thing is to disincentive lazy behaviour to be able to maximize confirmation rate.
```